modules/xmpp/strophe.jingle.js

```diff
@@ -22,6 +22,144 @@ const logger = getLogger(__filename);
 // function call chains.
 /* eslint-disable newline-per-chained-call */
 
+/**
+ * Reads a JSON-encoded message (from a "json-message" element) and extracts source descriptions. Adds the extracted
+ * source descriptions to the given Jingle IQ in the standard Jingle format.
+ *
+ * Encoding sources in this compact JSON format instead of standard Jingle was introduced in order to reduce the
+ * network traffic and load on the XMPP server. The format is described in Jicofo [TODO: insert link].
+ *
+ * @param {*} iq the IQ to which source descriptions will be added.
+ * @param {*} jsonMessageXml The XML node for the "json-message" element.
+ * @returns nothing.
+ */
+function expandSourcesFromJson(iq, jsonMessageXml) {
+
+    let json;
+    try {
+        json = JSON.parse(jsonMessageXml.textContent);
+    } catch (error) {
+        logger.error(`json-message XML contained invalid JSON, ignoring: ${jsonMessageXml.textContent}`);
+        return;
+    }
+
+    if (!json || !json.sources) {
+        // It might be a message of a different type, no need to log.
+        return;
+    }
+
+    // This is where we'll add "source" and "ssrc-group" elements. Create them elements if they don't exist.
+    let videoRtpDescription = getOrCreateRtpDescription(iq, 'video');
+    let audioRtpDescription = getOrCreateRtpDescription(iq, 'audio');
+
+    for (let owner in json.sources) {
+        let ownerSources = json.sources[owner];
+        // The video sources, video ssrc-groups, audio sources and audio ssrc-groups are encoded in that order in the
+        // elements of the array.
+        let videoSources = ownerSources && ownerSources.length > 0 && ownerSources[0];
+        let videoSsrcGroups = ownerSources && ownerSources.length > 1 && ownerSources[1];
+        let audioSources = ownerSources && ownerSources.length > 2 && ownerSources[2];
+        let audioSsrcGroups = ownerSources && ownerSources.length > 3 && ownerSources[3];
+
+        if (videoSources && videoSources.length > 0) {
+            for (let i = 0; i < videoSources.length; i++) {
+                videoRtpDescription.appendChild(createSourceExtension(owner, videoSources[i]))
+            }
+        }
+        if (videoSsrcGroups && videoSsrcGroups.length > 0) {
+            for (let i = 0; i < videoSsrcGroups.length; i++) {
+                videoRtpDescription.appendChild(createSsrcGroupExtension(videoSsrcGroups[i]))
+            }
+        }
+        if (audioSources && audioSources.length > 0) {
+            for (let i = 0; i < audioSources.length; i++) {
+                audioRtpDescription.appendChild(createSourceExtension(owner, audioSources[i]))
+            }
+        }
+        if (audioSsrcGroups && audioSsrcGroups.length > 0) {
+            for (let i = 0; i < audioSsrcGroups.length; i++) {
+                audioRtpDescription.appendChild(createSsrcGroupExtension(audioSsrcGroups[i]))
+            }
+        }
+    }
```

Comment:
Since we are already extracting all the source information from the json message, is it possible to log it here as a debug message so that we can get rid of the log that logs the full iq ?